File: cli/src/commands/read.rs
use std::io::Write;
use std::process::{Command, Stdio};

use crate::client::CowikiClient;
use crate::error::CliError;

pub async fn run(
    client: &CowikiClient,
    slug: &str,
    branch: &str,
    no_pager: bool,
    json: bool,
) -> Result<(), CliError> {
    let page = client.get_page(slug, branch).await.map_err(|e| match &e {
        CliError::Api { status: 404, .. } => {
            CliError::Unexpected(format!("page not found: \"{slug}\" on branch \"{branch}\""))
        }
        _ => e,
    })?;

    if json {
        println!(
            "{}",
            serde_json::to_string_pretty(&page)
                .map_err(|e| CliError::Unexpected(format!("failed to serialize page: {e}")))?
        );
        return Ok(());
    }

    let title_line = format!("\n  {}\n", colored::Colorize::bold(page.title.as_str()));
    let info_line = format!(
        "  slug: {} | branch: {}\n\n",
        page.slug, page.branch
    );

    let full_output = format!("{title_line}{info_line}{}", page.body);

    if no_pager {
        println!("{full_output}");
    } else {
        let pager = std::env::var("PAGER").unwrap_or_else(|_| "less -R".into());
        let pager_parts: Vec<&str> = pager.split_whitespace().collect();
        let (pager_cmd, pager_args) = pager_parts.split_first().unwrap_or((&"less", &["-R"][..]));

        let mut child = Command::new(pager_cmd)
            .args(pager_args)
            .stdin(Stdio::piped())
            .spawn()
            .map_err(|e| {
                CliError::Unexpected(format!("failed to spawn pager \"{pager}\": {e}"))
            })?;

        if let Some(mut stdin) = child.stdin.take() {
            stdin
                .write_all(full_output.as_bytes())
                .map_err(|e| CliError::Unexpected(format!("failed to write to pager: {e}")))?;
        }

        child
            .wait()
            .map_err(|e| CliError::Unexpected(format!("pager error: {e}")))?;
    }

    Ok(())
}

File: cli/src/commands/read.ts
import { Command } from 'commander';
import { spawn } from 'node:child_process';
import chalk from 'chalk';
import { loadConfig } from '../config.js';
import { CowikiClient } from '../client.js';
import { printError, printJson } from '../output.js';
import { resolveUserBranch, requireWorkspace } from '../shared.js';
import type { GlobalOpts } from '../shared.js';

export function registerReadCommand(program: Command): void {
  program
    .command('read')
    .description('Read a wiki page (supports --dir for wiki/entities/concepts)')
    .argument('<slug>', 'Page slug to read')
    .option('--no-pager', 'Print directly to stdout instead of using a pager')
    .option('--dir <dir>', 'Directory: wiki (default), entities, concepts, or subdir like wiki/messi')
    .action(async (slug, opts, cmd) => {
      const globalOpts = cmd.parent?.optsWithGlobals() as GlobalOpts;
      const config = loadConfig(globalOpts?.server)
      const ws = requireWorkspace(globalOpts.workspace);
      const client = new CowikiClient(config.baseUrl, config.apiKey);
      const branch = await resolveUserBranch(client);

      try {
        const page = await client.getPage(ws, slug, branch, opts.dir);

        if (globalOpts.json) {
          printJson(page);
          return;
        }

        const fullOutput =
          `\n  ${chalk.bold(page.title)}\n` +
          `  slug: ${page.slug} | branch: ${page.branch}\n\n` +
          page.body;

        if (opts.noPager) {
          console.log(fullOutput);
        } else {
          const pager = process.env.PAGER || 'less -R';
          const [pagerCmd, ...pagerArgs] = pager.split(/\s+/);

          const child = spawn(pagerCmd, pagerArgs, {
            stdio: ['pipe', 'inherit', 'inherit'],
          });

          child.stdin!.write(fullOutput);
          child.stdin!.end();

          await new Promise<void>((resolve, reject) => {
            child.on('close', (code) => {
              if (code === 0 || code === null) resolve();
              else reject(new Error(`pager exited with code ${code}`));
            });
            child.on('error', reject);
          });
        }
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        // Friendly 404
        if (msg.includes('404')) {
          printError(`page not found: "${slug}" on branch "${branch}"`);
        } else {
          printError(msg);
        }
        process.exit(1);
      }
    });
}
